### PR TITLE
/v1/client/allocation/./{stats,gc} ACL enforcement

### PR DESCRIPTION
**Pull request by schmichael**
The `/snapshots` endpoint will be covered in #3322

### CODE REVIEW

File: command/agent/alloc_endpoint.go [hunk 1]
```diff
@@ -118,6 +119,18 @@ func (s *HTTPServer) ClientGCRequest(resp http.ResponseWriter, req *http.Request
 }
 
 func (s *HTTPServer) allocGC(allocID string, resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	var secret string
+	s.parseToken(req, &secret)
+
+	var namespace string
+	parseNamespace(req, &namespace)
+
+	// Check namespace read-job permissions
```
Comment (chelseakomlo) [inline command/agent/alloc_endpoint.go]: Submit job permission

File: command/agent/alloc_endpoint_test.go [hunk 2]
```diff
@@ -243,6 +243,61 @@ func TestHTTP_AllocStats(t *testing.T) {
...
+			setToken(req, token)
+			_, err := s.Server.ClientAllocRequest(respW, req)
+			assert.NotNil(err)
+			assert.Contains(err.Error(), "unknown allocation ID")
+		}
+
+		// Try request with a management token
+		// Still returns an error because the alloc does not exist
+		{
+			respW := httptest.NewRecorder()
```
Comment (chelseakomlo) [inline command/agent/alloc_endpoint_test.go]: Maybe add an allocation and test the valid case?

Comment (dadgar) [reply command/agent/alloc_endpoint_test.go]: This and the one above is actually the valid case. He is making it pass the ACL section and hitting an error for it not existing. Would be more annoying to mock in a real alloc

**Comment by schmichael** [comment]
> Is a migrate token used in lieu of ACL permissions? I thought that both would be validated for the snapshot endpoint, as maybe there are any strange edge cases that happen with only validating one. But if there are cases I'm not thinking of, let me know!
-- @chelseakomlo 

Good question: I think `/snapshot` should support *either* the migration token or a standard ACL to allow operators to snapshot allocations manually (`namespace:submit-job` maybe?).

I didn't want to touch that endpoint in this PR to avoid merge conflicts, so let's figure it out in your PR.

**Pull request merged**